vertical-pod-autoscaler/updater/eviction/pods_eviction_restriction.go

```diff
@@ -0,0 +1,233 @@
+// Does not check if pod was actually evicted after eviction grace period.
+func (e *podsEvictionRestrictionImpl) Evict(podToEvict *apiv1.Pod) error {
+	cr, present := e.podsCreators[getPodId(podToEvict)]
+	if !present {
+		return fmt.Errorf("pod not suitable for eviction %v : not in replicated pods map", podToEvict.Name)
+	}
+	if e.evictionBudget[cr] < 1 {
+		return fmt.Errorf("cannot evict pod %v : eviction budget exceeded", podToEvict.Name)
+	}
+
+	eviction := &policyv1.Eviction{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: podToEvict.Namespace,
+			Name:      podToEvict.Name,
+		},
+	}
+	err := e.client.CoreV1().Pods(podToEvict.Namespace).Evict(eviction)
+	if err != nil {
+		glog.Errorf("failed to evict pod %s, error: %v", podToEvict.Name, err)
+		return err
+	}
+	e.evictionBudget[cr] = e.evictionBudget[cr] - 1
+	return nil
+}
+
+// NewPodsEvictionRestrictionFactory creates PodsEvictionRestrictionFactory
+func NewPodsEvictionRestrictionFactory(client kube_client.Interface, minReplicas int, evictionToleranceFraction float64) PodsEvictionRestrictionFactory {
+	return &podsEvictionRestrictionFactoryImpl{client: client, minReplicas: minReplicas, evictionToleranceFraction: evictionToleranceFraction}
+}
+
+// NewPodsEvictionRestriction creates PodsEvictionRestriction for given set of pods.
+func (f *podsEvictionRestrictionFactoryImpl) NewPodsEvictionRestriction(pods []*apiv1.Pod) PodsEvictionRestriction {
+	// We can evict pod only if it is a part of replica set
+	// For each replica set we can evict only a fraction of pods.
+	// Evictions may be later limited by pod disruption budget if configured.
+
+	livePods := make(map[podReplicaCreator][]*apiv1.Pod)
+
+	for _, pod := range pods {
+		creator, err := getPodReplicaCreator(pod)
+		if err != nil {
+			glog.Errorf("failed to obtain replication info for pod: %v", pod.Name, err)
+			continue
+		}
+		if creator == nil {
+			glog.V(1).Infof("pod not replicated %v", pod.Name)
+			continue
+		}
+		livePods[*creator] = append(livePods[*creator], pod)
+	}
+
+	podsCreators := make(map[string]podReplicaCreator)
+	creatorsEvictionBudget := make(map[podReplicaCreator]int)
+	for creator, replicas := range livePods {
+		actual := len(replicas)
+		if actual < f.minReplicas {
+			glog.V(1).Infof("too few replicas for %v %v/%v. Found %v live pods",
+				creator.Kind, creator.Namespace, creator.Name, actual)
+			continue
+		}
+		var configured int
+		if creator.Kind == "Job" {
+			// Job has no replicas configuration, so we will use actual number of live pods as replicas count.
+			configured = actual
+		} else {
+			var err error
+			configured, err = getReplicaCount(creator, f.client)
+			if err != nil {
+				glog.Errorf("failed to obtain replication info for %v %v/%v. %v",
+					creator.Kind, creator.Namespace, creator.Name, err)
+				continue
+			}
+		}
+
+		evictionTolerance := int(float64(configured) * f.evictionToleranceFraction)
```

Comment:
round it up

Reply:
I'd really like to understand what your intention was here, @mwielgus . 

The code was changed to allow 1 pod eviction in each group of at least minReplicas without any further conditions. Which in particular means if we're causing the pods to be unschedulable, we'll eventually destroy the whole group, taking out X pods in first update attempt and then 1 pod in each subsequent updater run (even though there is growing number of unscheduled pods).

I think what you wanted instead was to set evictionTolerance to 1 if it's smaller than that (but still respect currentlyEvicted), right?

Reply:
It was almost a year ago but AFAIR the reason was not to eat the whole PDB